### ISSUE TITLE
Non-blocking Socket and Early Data

### ISSUE THREAD

**Issue by pracj3am** [openssl/openssl#7757]
- OpenSSL version: 1.1.1a
  - [Configuration data](https://github.com/openssl/openssl/files/2644168/configdata.txt)
  - OS Name, Version, Hardware platform: Debian Stretch on x64
  - Application Details: nginx 1.15.6

Here is a excerpt from nginx log with additional debug messages:
```
2018/12/04 13:07:13 [debug] 18272#0: *124 SSL ALPN supported by client: h2
2018/12/04 13:07:13 [debug] 18272#0: *124 SSL ALPN supported by client: http/1.1
2018/12/04 13:07:13 [debug] 18272#0: *124 SSL ALPN selected: h2
2018/12/04 13:07:13 [debug] 18272#0: *124 SSL_read_early_data: 1, 1
2018/12/04 13:07:13 [debug] 18272#0: *124 SSL_get_current_cipher()
2018/12/04 13:07:13 [debug] 18272#0: *124 SSL: TLSv1.3, cipher: "TLS_AES_256_GCM_SHA384 TLSv1.3 Kx=any Au=any Enc=AESGCM(256) Mac=AEAD"
2018/12/04 13:07:13 [debug] 18272#0: *124 SSL_session_reused()
2018/12/04 13:07:13 [debug] 18272#0: *124 SSL reused session
[...]
2018/12/04 13:07:13 [debug] 18272#0: *125 SSL_read_early_data: 1, 529
2018/12/04 13:07:13 [debug] 18272#0: *125 SSL_read_early_data: 1, 72
2018/12/04 13:07:13 [debug] 18272#0: *125 SSL_read_early_data: 1, 61
2018/12/04 13:07:13 [debug] 18272#0: *125 SSL_read_early_data: 1, 68
2018/12/04 13:07:13 [debug] 18272#0: *125 SSL_read_early_data: 1, 63
2018/12/04 13:07:13 [debug] 18272#0: *125 SSL_read_early_data: 0, 0
2018/12/04 13:07:13 [debug] 18272#0: *125 SSL_get_error: 2
[...]
2018/12/04 13:07:13 [debug] 18272#0: *125 SSL to write: 344
2018/12/04 13:07:13 [debug] 18272#0: *125 SSL_write_early_data: 1, 344
[...]
2018/12/04 13:07:13 [debug] 18272#0: *125 SSL to write: 16384
2018/12/04 13:07:13 [debug] 18272#0: *125 SSL_write_early_data: 1, 16384
2018/12/04 13:07:13 [debug] 18272#0: *125 SSL to write: 16384
2018/12/04 13:07:13 [debug] 18272#0: *125 SSL_write_early_data: 0, 0
2018/12/04 13:07:13 [debug] 18272#0: *125 SSL_get_error: 3
[...]
2018/12/04 13:07:13 [debug] 18272#0: *125 ssl new session: 915F34E7:32:157
2018/12/04 13:07:13 [debug] 18272#0: *125 SSL_read_early_data: 2, 0
2018/12/04 13:07:13 [alert] 18272#0: *125 ignoring stale global SSL error (at 6) (SSL: error:1409E10F:SSL routines:ssl3_write_bytes:bad length:len:257, rlayer.wnum:0, wb->left:2882, rlayer.wpend_tot:16384, early_data_state:12
) while processing HTTP/2 connection
2018/12/04 13:07:13 [debug] 18272#0: *125 SSL_read: -1
2018/12/04 13:07:13 [debug] 18272#0: *125 SSL_get_error: 5
```

The numbers after the `SSL_read_early_data` is a return value and readbytes. The numbers after the `SSL_write_early_data` is a return value and written bytes.

### My Interpretation

Call to the `SSL_write_early_data` returns 0 with SSL error `SSL_ERROR_WANT_WRITE` and so we wait until an underlying event library notifies us that the fd is ready for writing to retry the write. In the meantime a new SSL session is created and `ssl3_handshake_write` function is called. It calls `ssl3_write_bytes`, where the following check does not pass, because of shorter length
```c
    /*
     * ensure that if we end up with a smaller value of data to write out
     * than the original len from a write which didn't complete for
     * non-blocking I/O and also somehow ended up avoiding the check for
     * this in ssl3_write_pending/SSL_R_BAD_WRITE_RETRY as it must never be
     * possible to end up with (len-tot) as a large number that will then
     * promptly send beyond the end of the users buffer ... so we trap and
     * report the error in a way the user will notice
     */
    if ((len < s->rlayer.wnum)
        || ((wb->left != 0) && (len < (s->rlayer.wnum + s->rlayer.wpend_tot)))) {
        SSLfatal(s, SSL_AD_INTERNAL_ERROR, SSL_F_SSL3_WRITE_BYTES,
                 SSL_R_BAD_LENGTH);
    }
```
Than the end of early data record arrives which in the nginx code triggers calling `ERR_peek_error` and then `SSL_read` which fails.

**Comment by mattcaswell** [comment]
Yes, your analysis looks plausible. Are you able to reproduce this error - or was this a one off?

**Comment by mattcaswell** [comment]
I'm thinking that this not should even be a valid thing to try and do, i.e. call SSL_read_early_data() where there is an outstanding unsuccessful SSL_write_early_data() pending.

There is this statement on the `SSL_get_error()` page which disallows this for SSL_read()/SSL_write():
````
It is safe to call SSL_read() or SSL_read_ex() when more data is available
even when the call that set this error was an SSL_write() or SSL_write_ex().
However if the call was an SSL_write() or SSL_write_ex(), it should be called
again to continue sending the application data.
````
Since SSL_read_early_data()/SSL_write_early_data() basically call SSL_read()/SSL_write() internally that statement should probably be extended to cover the early data forms.

So, probably this is a documentation issue in OpenSSL and nginx needs to be fixed to not do this.

**Comment by davidben** [comment]
Being able to read and write in parallel is quite important to an HTTP/2 implementation. It certainly used to work before for SSL_read/SSL_write.

0-RTT makes this a little bit tricky because of read-triggered handshake data, but I think the TLS stack should make that work. This was a case we made sure to think through when designing our 0-RTT API and implementation. On the server, it's possible to completely avoid this scenario with half-RTT tickets. On the client, the scenario is unavoidable but possible to abstract away.

**Comment by mattcaswell** [comment]
> It certainly used to work before for SSL_read/SSL_write.

AFAIK this has never worked for SSL_read/SSL_write. The restriction was always there - we just documented it. Most of the time it works anyway but there are cases (e.g. with reneg) where it doesn't.

**Comment by davidben** [comment]
(HTTP/2 implementations are supposed to disable renegotiation. Though it's true most don't since OpenSSL didn't provide an API for it until recently.)

I believe renegotiation would just cause `SSL_read` and `SSL_write` to both drive the handshake, which works mostly fine, modulo the error idempotence issue (the mess surrounding CVE-2017-3737). But even then, that "mostly working", renegotiation being uncommon, and concurrent read and write being a very common application need, I do not think it's reasonable restriction for OpenSSL to now impose. Indeed this interaction with 0-RTT and renegotiation was one of the main reasons we decided to fully support error idempotence on our end by latching the errors. Go's crypto/tls similarly latches errors.

But I don't think that's even what the current text forbids. It says:

> It is safe to call SSL_read() or SSL_read_ex() when more data is available even when the call that set this error was an SSL_write() or SSL_write_ex().

This says that SSL_read *is* okay to call when SSL_write wants data.

> However if the call was an SSL_write() or SSL_write_ex(), it should be called again to continue sending the application data."

This is a little confusing because "the call" could refer to a couple things. It reads to me that, if SSL_write blocked on write, one must repeat the same SSL_write call again to continue. That is true, I don't think that's what's going on here. I suspect the internal check for the second sentence is breaking the first sentence. (We reworked writes for handshake data to account for this in 0-RTT.)

**Comment by pracj3am** [comment]
> Are you able to reproduce this error - or was this a one off?
 
Yes, I am able to reproduce it with 100%-rate on certain website in Firefox, if early data and HTTP/2 is turned on. The result is that no stylesheets or javascript are loaded.

**Comment by mattcaswell** [comment]
> This says that SSL_read is okay to call when SSL_write wants data.

Correct

>It reads to me that, if SSL_write blocked on write, one must repeat the same SSL_write call again to continue.

Also correct - that's what it means.

**Comment by mattcaswell** [comment]
i.e. if SSL_write is blocked on write you can't call SSL_read. You must call SSL_write again.

**Comment by davidben** [comment]
Sorry, my text was even more confusing than the original text. :-)

> i.e. if SSL_write is blocked on write you can't call SSL_read. You must call SSL_write again.

The text only says that it should be called again "to continue sending the application data". It doesn't say you cannot call other stuff. If the intent was indeed to block out everything, I think that was solidly a mistake. (E.g. NGINX does it a lot.) At the very least, you all should enforce it more consistently: make SSL_read fail if there is pending write data, rather than only flakily failing when hitting a read-triggered write. If you do that, I think you will very quickly find that lots of clients break this assumption.

It also contradicts the first sentence, unless you read "this error" to mean SSL_ERROR_WANT_READ. But I don't think the text supports that. If anything, it's in the SSL_ERROR_WANT_WRITE (In general, this text uses "it", "this", and "the call" too much. It's very hard to understand.)

**Comment by kroeckx** [comment]
> i.e. if SSL_write is blocked on write you can't call SSL_read. You must call SSL_write again.

The text was supposed to make clear that you can, but clearly
the text is not clear enough.

If SSL_write() fails, and you call it again at some point later,
the length should be the same. But you can call SSL_read() between
those SSL_write() calls.

If that's not the case, I think it's a bug in OpenSSL.

**Comment by mattcaswell** [comment]
Ok. Lets take a step back and describe the problem as I see it with interleaving SSL_read()/SSL_write(). We can then look to see if there is a sane solution.

SSL_write is supposed to be an atomic operation (assuming you don't set the non default option SSL_MODE_ENABLE_PARTIAL_WRITE), i.e. it either completely succeeds with everything written (positive return value) or it completely fails (SSL_ERROR_WANT_WRITE or similar) with nothing written (at least nothing is reported as written to the user application). Technically there is also the option of a fatal error - in which case we are in an indeterminate middle state. In practice of course the underlying network is not atomic so what it actually does is split the data up into records and writes each one in turn. If anything fails (SSL_ERROR_WANT_WRITE) control is passed back to the application which must call SSL_write again with exactly the same data. As far as the user application is concerned nothing has yet been wrriten - but in practice it may have done. The user application has to keep calling SSL_write() until everything has been successfully written.

Consider the scenario of a large write which needs to be broken up into 3 records. Lets say the first record gets written to the network successfully. During processing the 2nd record libssl creates the record, encrypts it and puts it in its buffer ready to write out. The writing fails (SSL_ERROR_WANT_WRITE). At this point we have some data already written to the network, some data already encrypted and buffered but not yet written, and some data still to process. Control is passed back to the user application.

Now lets say a record arrives for reading. Instead of continuing with the in progress write operation the application calls SSL_read. The read packet is a reneg ClientHello so a new handshake is initiated and the SSL_read call will actually attempt to *write* out a ServerHello. At this point we have a problem because the write buffers are already full and contain an encrypted application data record ready to write out.

In this scenario what is the correct behaviour for SSL_read? Possible options:

1) Ditch the buffered write data. This would work in that we would then be able to proceed with writing out the ServerHello. However as far as the application is concerned it is still part way through an ongoing SSL_write operation which it will then attempt to continue - and libssl would have to re-encrypt the second record and carry on from where it left off. But we've had a reneg occur since we started the write operation - so that means that the first record we wrote happened before the reneg, and subsequent records will be written after the reneg. This sounds like a really "bad idea". It would be totally unexpected, and (potentially) dangerous.

2) Flush the buffered write data. Again this would work but suffers from the same flaw as (1) above. In this case we would have written the first 2 records before the reneg and the 3rd record would be after the reneg. Again this sounds like a bad idea.

3) Write *all* the pending application data before writing the ServerHello. However we've only buffered the second record - so in order for this option to work we would have to suck in the *entire* SSL_write data that the application presented to us when it first called SSL_write and buffer it somewhere. This could potentially be very large and means lots of copying data around. We probably don't want to do this.

4) SSL_read could fail and signal that the user application needs to complete the SSL_write operation first. But we don't have an error code for this. All the SSL_ERROR_WANT_* codes mean "call SSL_read again", not "stop calling SSL_read and instead call SSL_write". We could introduce a new one or redefine SSL_ERROR_WANT_WRITE to mean "call SSL_read again unless there is a pending SSL_write operation which also returned SSL_ERROR_WANT_WRITE in which case call that" - but this option doesn't help existing applications and would require those applications to change to be able to cope with the new code or redefined code.

Where I thought we were at with this is documenting that you shouldn't call SSL_read if there is an ongoing SSL_write operation. If we don't consider that viable then I think the only sane option is (4) - but that will still require applications that are running into this problem to make a change.

**Comment by kroeckx** [comment]
SSL_read() should return SSL_ERROR_WANT_WRITE in that case. It's
already documented that SSL_read() can return
SSL_ERROR_WANT_WRITE.

If SSL_read() returns SSL_ERROR_WANT_WRITE, and there is an SSL_write()
that didn't send all data, calling SSL_read() should continue to return
SSL_ERROR_WANT_WRITE as long as the data from the SSL_write and the
renegiotation after it is not written.

That means that we should not try to write internally until at
least all data from the SSL_write() has been encrypted and is
at least queued.

**Comment by kroeckx** [comment]
Note that after SSL_write() is done with the user data, it should
pick up were SSL_read() stopped, otherwise applications that use
select() will most likely get stuck since they have no reason to
try and read.

**Comment by davidben** [comment]
I'm not sure that'll work for the same reason that (3) won't work. SSL_write does not have a way to return the output to the caller.

**Comment by kroeckx** [comment]
If we internally send something, we have an internal buffer to do
so, and that's used again if it failed.

**Comment by t8m** [comment]
@kroeckx @mattcaswell Is this still an issue with currently supported releases? I assume it is.

**Comment by t8m** [comment]
@mattcaswell We should consider whether this makes sense to implement (as a feature) or rather not.
